Duplicated Validators in Configuration

If the configuration file used is

~~~
<redpen-conf lang="en" variant="gb">
    <validators>
	<validator name="Spelling">
            <property name="dict" value="redpen-spelling-dict"/>
        </validator>
	<validator name="Spelling" \>
</redpen-conf>
~~~

the Spelling validator will be run twice.

~~~
$ cat foo.txt 
speling
$ redpen -f markdown foo.txt
[2017-04-13 17:26:50.777][INFO ] cc.redpen.Main - Configuration file: /tmp/redpen-conf-en.xml
[2017-04-13 17:26:50.782][INFO ] cc.redpen.config.ConfigurationLoader - Loading config from specified config file: "/tmp/redpen-conf-en.xml"
[2017-04-13 17:26:50.789][INFO ] cc.redpen.config.ConfigurationLoader - Succeeded to load configuration file
[2017-04-13 17:26:50.789][INFO ] cc.redpen.config.ConfigurationLoader - Language is set to "en"
[2017-04-13 17:26:50.789][INFO ] cc.redpen.config.ConfigurationLoader - Variant is set to "gb"
[2017-04-13 17:26:50.790][INFO ] cc.redpen.config.ConfigurationLoader - No "symbols" block found in the configuration
[2017-04-13 17:26:50.793][INFO ] cc.redpen.config.SymbolTable - Default symbol settings are loaded
[2017-04-13 17:26:50.861][INFO ] cc.redpen.parser.SentenceExtractor - "[., ?, !]" are added as a end of sentence characters
[2017-04-13 17:26:50.861][INFO ] cc.redpen.parser.SentenceExtractor - "[', "]" are added as a right quotation characters
[2017-04-13 17:26:51.113][INFO ] org.reflections.Reflections - Reflections took 32 ms to scan 1 urls, producing 5 keys and 53 values 
[2017-04-13 17:26:51.167][WARN ] cc.redpen.validator.ValidatorFactory - cc.redpen.validator.sentence.SpaceBeginningOfSentenceValidator is deprecated
[2017-04-13 17:26:51.185][WARN ] cc.redpen.validator.ValidatorFactory - cc.redpen.validator.section.VoidSectionValidator is deprecated
[2017-04-13 17:26:51.202][INFO ] org.reflections.Reflections - Reflections took 3 ms to scan 1 urls, producing 160 keys and 163 values 
[2017-04-13 17:26:51.285][INFO ] cc.redpen.util.DictionaryLoader - Succeeded to load SpellingValidator default dictionary.
[2017-04-13 17:26:51.289][INFO ] cc.redpen.util.DictionaryLoader - Succeeded to load SpellingValidator user dictionary.
foo.txt:1: ValidationError[Spelling], Found possibly misspelled word "speling". at line: speling
foo.txt:1: ValidationError[Spelling], Found possibly misspelled word "speling". at line: speling

[2017-04-13 17:26:51.299][ERROR] cc.redpen.Main - The number of errors "2" is larger than specified (limit is "1").
~~~

The error at the configuration file can be difficult to find. Would be nice to have a warning saying that the validator is duplicated and possible only using the last definition.

--- a/redpen-core/src/main/java/cc/redpen/config/ConfigurationLoader.java
+++ b/redpen-core/src/main/java/cc/redpen/config/ConfigurationLoader.java
@@ -32,6 +32,8 @@ import javax.xml.parsers.DocumentBuilderFactory;
 import javax.xml.parsers.ParserConfigurationException;
 import java.io.*;
 import java.nio.charset.StandardCharsets;
+import java.util.LinkedHashMap;
+import java.util.Map;
 
 import static cc.redpen.config.Configuration.ConfigurationBuilder;
 
@@ -214,6 +216,7 @@ public class ConfigurationLoader {
 
     private void extractValidatorConfigurations(ConfigurationBuilder configBuilder, NodeList validatorElementList) {
         ValidatorConfiguration currentConfiguration;
+        Map<String, ValidatorConfiguration> validatorConfigurations = new LinkedHashMap<>();
         for (int i = 0; i < validatorElementList.getLength(); i++) {
             Node nNode = validatorElementList.item(i);
             if (nNode.getNodeType() != Node.ELEMENT_NODE) {
@@ -221,9 +224,13 @@ public class ConfigurationLoader {
             }
             Element element = (Element) nNode;
             if (element.getNodeName().equals("validator")) {
+                final String currentValidatorName = element.getAttribute("name");
                 currentConfiguration =
-                        new ValidatorConfiguration(element.getAttribute("name"));
-                configBuilder.addValidatorConfig(currentConfiguration);
+                        new ValidatorConfiguration(currentValidatorName);
+                if (validatorConfigurations.containsKey(currentValidatorName)) {
+                    LOG.warn("Duplicated validator configuration was found: " + currentValidatorName);
+                }
+                validatorConfigurations.put(currentValidatorName, currentConfiguration);
                 NodeList propertyElementList = nNode.getChildNodes();
                 extractProperties(currentConfiguration, propertyElementList);
             } else {
@@ -231,6 +238,7 @@ public class ConfigurationLoader {
                 LOG.warn("Skip this block ...");
             }
         }
+        validatorConfigurations.values().forEach(configBuilder::addValidatorConfig);
     }
 
     private void extractProperties(ValidatorConfiguration currentConfiguration,

--- a/redpen-core/src/test/java/cc/redpen/config/ConfigurationLoaderTest.java
+++ b/redpen-core/src/test/java/cc/redpen/config/ConfigurationLoaderTest.java
@@ -408,4 +408,31 @@ public class ConfigurationLoaderTest {
         Configuration config = new ConfigurationLoader().load(file);
         assertEquals(file.getParentFile(), config.getBase());
     }
+
+    @Test
+    public void testDuplicatedValidatorConfiguration() throws RedPenException{
+        String sampleConfigString =
+                "<redpen-conf lang=\"ja\">" +
+                        "<validators>" +
+                        "<validator name=\"SentenceLength\">" +
+                        "<property name=\"max_length\" value=\"200\" />" +
+                        "</validator>" +
+                        "<validator name=\"SentenceLength\">" +
+                        "<property name=\"max_length\" value=\"300\" />" +
+                        "</validator>" +
+                        "</validators>" +
+                        "<symbols>" +
+                        "<symbol name=\"EXCLAMATION_MARK\" value=\"！\" invalid-chars=\"!\" after-space=\"true\" />" +
+                        "</symbols>" +
+                        "</redpen-conf>";
+
+        Configuration configuration = new ConfigurationLoader().loadFromString(sampleConfigString);
+
+        assertNotNull(configuration);
+        assertEquals(1, configuration.getValidatorConfigs().size());
+        assertEquals("SentenceLength",
+                configuration.getValidatorConfigs().get(0).getConfigurationName());
+        assertEquals("300",
+                configuration.getValidatorConfigs().get(0).getProperty("max_length"));
+    }
 }

